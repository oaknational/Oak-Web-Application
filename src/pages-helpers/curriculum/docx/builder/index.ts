export { default as frontCover } from "./1_frontCover";
export { default as tableOfContents } from "./2_tableOfContents";
export { default as ourCurriculum } from "./3_ourCurriculum";
export { default as threadsExplainer } from "./4_threadsExplainer";
export { default as subjectExplainer } from "./5_subjectExplainer";
export { default as subjectPrincipals } from "./6_subjectPrincipals";
export { default as ourPartner } from "./7_ourPartner";
export { default as units } from "./8_units";
export { default as threadsOverview } from "./9_threadsOverview";
export { default as threadsDetail } from "./10_threadsDetail";
export { default as backCover } from "./11_backCover";
export { default as pageLayout } from "./12_pageLayout";
