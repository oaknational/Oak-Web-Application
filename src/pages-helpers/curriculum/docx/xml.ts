import { js2xml, xml2js } from "xml-js";
import type { Element } from "xml-js";

export function xmlRootToJson(xmlData: string) {
  return xml2js(xmlData, {
    compact: false,
    captureSpacesBetweenElements: false,
  });
}

/**
 * To support <https://prettier.io/docs/en/options#embedded-language-formatting> as well as assertion.
 *
 * NOTE: This must called `xml`
 *
 * @param strings template strings
 * @param values tamplate values
 * @returns xml string
 */
export function safeXml(
  strings: TemplateStringsArray,
  ...values: (string | number | string[])[]
) {
  let outXml = "";
  for (let i = 0; i < strings.length; i++) {
    if (values[i] !== undefined) {
      const valueRaw = values[i];
      const value = Array.isArray(valueRaw) ? valueRaw.join("") : valueRaw;
      outXml += `${strings[i]!}${value}`;
    } else {
      outXml += `${strings[i]!}`;
    }
  }

  // Just for assertion
  xmlRootToJson(outXml);

  return outXml;
}

export function xmlElementToJson(xmlData?: string) {
  if (!xmlData) return;
  const rootNode = xmlRootToJson(xmlData);
  if (rootNode.elements.length !== 1) {
    throw new Error("Expecting a single root node in XML");
  }
  return collapseFragments(rootNode as Element)?.elements?.[0];
}

export function createFragment(elements: Element[]): Element {
  return {
    type: "element",
    name: "XML_FRAGMENT",
    elements: elements,
  };
}

export function collapseFragments(root: Element) {
  const run = (orig: Element) => {
    let node = orig;
    const cloneIfRequired = () => {
      if (node === orig) {
        return {
          ...node,
          elements: node.elements ? [...node.elements] : undefined,
        };
      }
      return node;
    };
    if (node.elements) {
      for (let i = node.elements.length - 1; i >= 0; i--) {
        if (node.elements) {
          const child = node.elements[i]!;
          const out = run(child);

          if (Array.isArray(out)) {
            node = cloneIfRequired();
            node.elements!.splice(i, 1, ...out);
          } else if (out !== child) {
            node = cloneIfRequired();
            if (node.elements) {
              node.elements[i] = out;
            }
          }
        }
      }
    }

    if (node.name === "XML_FRAGMENT") {
      return node.elements ?? [];
    }
    return node;
  };

  const out = run(root);
  if (Array.isArray(out)) {
    return createFragment(out);
  }
  return out;
}

export function jsonXmlToXmlString(json: Element): string {
  const root = "name" in json ? { elements: [json] } : json;
  const output = js2xml(collapseFragments(root));
  return output;
}

export function cdata(input: string | number) {
  return `<![CDATA[${input}]]>`;
}

export function cdataJson(input: Element): Element {
  if (input.type !== "text") {
    throw new Error("Expecting text node");
  }
  return {
    type: "cdata",
    cdata: String(input.text),
  };
}
