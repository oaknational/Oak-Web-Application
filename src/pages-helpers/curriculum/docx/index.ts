import { join } from "path";

import { format } from "date-fns";

import type {
  CurriculumOverviewMVData,
  CurriculumUnitsTabData,
} from "../../../node-lib/curriculum-api-2023";
import type { CurriculumOverviewSanityData } from "../../../common-lib/cms-types";

import * as builder from "./builder";
import {
  cmToEmu,
  cmToTwip,
  createImage,
  generateEmptyDocx,
  insertFooters,
  insertImages,
} from "./docx";

import { Ks4Option } from "@/node-lib/curriculum-api-2023/queries/curriculumPhaseOptions/curriculumPhaseOptions.schema";

export type CombinedCurriculumData = CurriculumUnitsTabData &
  CurriculumOverviewMVData &
  CurriculumOverviewSanityData;

export type Slugs = {
  subjectSlug: string;
  phaseSlug?: string;
  ks4OptionSlug?: string;
  keyStageSlug?: string;
  tierSlug?: string;
  childSubjectSlug?: string;
};

const ENABLE_TIMER = false;
const measure = (key: string, fn: () => void | Promise<void>) => {
  if (!ENABLE_TIMER) {
    const v = fn();
    return v;
  }
  let err: unknown;
  let v;
  console.time(key);
  try {
    v = fn();
  } catch (_err: unknown) {
    err = _err;
  } finally {
    console.timeEnd(key);
  }

  if (err) throw err;
  return v;
};

export default async function docx(
  data: CombinedCurriculumData,
  slugs: Slugs,
  ks4Options: Ks4Option[],
) {
  const zip = await generateEmptyDocx();

  const images = await insertImages(
    zip,
    {
      footerImage: join(
        process.cwd(),
        "src/pages-helpers/curriculum/docx/builder/images/footer-logo.png",
      ),
    },
    "footer-default.xml",
  );

  const footers = await insertFooters(zip, {
    default: `
      <w:p>
        <w:pPr>
            <w:jc w:val="right"/>
        </w:pPr>
        <w:r>
            <w:rPr>
                  <w:rFonts
                    w:ascii="Arial"
                    w:eastAsia="Arial"
                    w:hAnsi="Arial"
                    w:cs="Arial"
                  />
                  <w:rtl w:val="0"/>
            </w:rPr>
            <w:t xml:space="preserve">Exported ${format(
              new Date(),
              "dd MMMM yyyy",
            )}     </w:t>
            <w:tab/>
        </w:r>
        <w:r>
            <w:rPr>
              <w:rFonts
                w:ascii="Arial"
                w:eastAsia="Arial"
                w:hAnsi="Arial"
                w:cs="Arial"
              />
            </w:rPr>
            <w:fldChar w:fldCharType="begin"/>
            <w:instrText xml:space="preserve">PAGE</w:instrText>
            <w:fldChar w:fldCharType="separate"/>
            <w:fldChar w:fldCharType="end"/>
        </w:r>
        <w:r>
            <w:rPr>
                <w:rtl w:val="0"/>
            </w:rPr>
        </w:r>
        <w:r>
            ${createImage(images.footerImage, {
              width: cmToEmu(5.33),
              height: cmToEmu(0.55),
              xPos: cmToEmu(0.001),
              yPos: cmToEmu(0.001),
              xPosAnchor: "margin",
              yPosAnchor: "line",
              isDecorative: true,
            })}
        </w:r>
    </w:p>
    `,
  });

  // Run through the builders
  const runners = {
    frontCover: async () => await builder.frontCover(zip, { data, slugs }),
    frontCoverPageLayout: async () =>
      await builder.pageLayout(zip, {
        margins: {
          top: cmToTwip(1.5),
          right: cmToTwip(1.5),
          bottom: cmToTwip(1.5),
          left: cmToTwip(1.5),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    tableOfContents: async () =>
      await builder.tableOfContents(zip, { data, ks4Options }),
    tableOfContentsPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.25),
          right: cmToTwip(1.25),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.25),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    ourCurriculum: async () => await builder.ourCurriculum(zip),
    threadsExplainer: async () =>
      await builder.threadsExplainer(zip, { slugs, data }),
    subjectExplainer: async () => await builder.subjectExplainer(zip, { data }),
    subjectPrincipals: async () => await builder.subjectPrincipals(),
    ourPartner: async () => await builder.ourPartner(zip, { data }),
    units: async () => await builder.units(zip, { data, slugs, ks4Options }),
    threadsOverview: async () => await builder.threadsOverview(zip, { data }),
    threadsOverviewPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.5),
          right: cmToTwip(1.5),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.5),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    threadsDetail: async () => await builder.threadsDetail(zip, { data }),
    threadsDetailPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        margins: {
          top: cmToTwip(1.75),
          right: cmToTwip(1.75),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.75),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
    backCover: async () => await builder.backCover(zip, { data }),
    backCoverPageLayout: async () =>
      await builder.pageLayout(zip, {
        footers,
        isLast: true,
        margins: {
          top: cmToTwip(1.25),
          right: cmToTwip(1.25),
          bottom: cmToTwip(2.5),
          left: cmToTwip(1.25),
          header: cmToTwip(1.5),
          footer: cmToTwip(1.5),
        },
      }),
  };

  // Note: Just some code to messure some timings.
  await measure("total", async () => {
    for (const [key, runner] of Object.entries(runners)) {
      await measure(key, async () => {
        await runner();
      });
    }
  });

  return await zip.zipToBuffer();
}
