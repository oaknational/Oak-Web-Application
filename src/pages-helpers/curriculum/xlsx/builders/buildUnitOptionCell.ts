import { cartesianToExcelCoords } from "@ooxml-tools/units";
import { cdata, safeXml } from "@ooxml-tools/xml";

import { UnitOption } from "@/utils/curriculum/types";

export function buildUnitOptionCell<T extends Record<string, string>>(
  cellStyleIndexMap: T,
  x: number,
  y: number,
  unitOption: UnitOption,
  unitOptionIndex: number,
) {
  return safeXml`
    <c
      r="${cartesianToExcelCoords([x, y])}"
      t="inlineStr"
      s="${cellStyleIndexMap.temp3!}"
    >
      <is>
        <r>
          <rPr>
            <sz val="14" />
            <color rgb="FF000000" />
            <rFont val="Arial" />
            <family val="2" />
          </rPr>
          <t xml:space="preserve">${cdata(
              `\nOption ${unitOptionIndex + 1}\n`,
            )}</t>
        </r>
        <r>
          <rPr>
            <sz val="14" />
            <color rgb="FF000000" />
            <rFont val="Arial Bold" />
            <family val="2" />
          </rPr>
          <t xml:space="preserve">${cdata(`${unitOption.title}\n`)}</t>
        </r>
      </is>
    </c>
  `;
}
