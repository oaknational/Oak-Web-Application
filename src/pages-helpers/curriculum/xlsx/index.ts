import { cartesianToExcelCoords, pxToColumnWidth } from "@ooxml-tools/units";

import { generateEmptyXlsx, JSZipCached } from "../docx/docx";
import { cdata, safeXml, xmlCompact } from "../docx/xml";
import { formatCurriculumUnitsData } from "../docx/tab-helpers";

import { createXmlIndexMap } from "./builder";

import { Unit } from "@/utils/curriculum/types";
import {
  CurriculumOverviewMVData,
  CurriculumUnitsTabData,
} from "@/node-lib/curriculum-api-2023";

export type Slugs = {
  subjectSlug: string;
  phaseSlug?: string;
  ks4OptionSlug?: string;
  keyStageSlug?: string;
  tierSlug?: string;
  childSubjectSlug?: string;
};

function addOrUpdateSheet(zip: JSZipCached, sheetNumber: number, xml: string) {
  zip.writeString(
    `xl/worksheets/sheet${sheetNumber}.xml`,
    `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>${xml}`,
  );
}

function buildStyle() {
  const { xml: fontsXml, indexMap: fontIndexMap } = createXmlIndexMap({
    arial12: safeXml`
      <font>
        <sz val="12" />
        <color theme="1" />
        <name val="Arial" />
        <family val="2" />
      </font>
    `,
    arialBold22: safeXml`
      <font>
        <sz val="22" />
        <color rgb="FFFFFFFF" />
        <name val="Arial Bold" />
        <family val="2" />
      </font>
    `,
    arialBold24: safeXml`
      <font>
        <sz val="24" />
        <color rgb="FF000000" />
        <name val="Arial Bold" />
        <family val="2" />
      </font>
    `,
    arialBold14: safeXml`
      <font>
        <sz val="14" />
        <color rgb="FF000000" />
        <name val="Arial Bold" />
        <family val="2" />
      </font>
    `,
  });

  const { xml: fillsXml, indexMap: fillIndexMap } = createXmlIndexMap({
    none: safeXml`
      <fill>
        <patternFill patternType="none" />
      </fill>
    `,
    temp1: safeXml`
      <fill>
        <patternFill patternType="gray125" />
      </fill>
    `,
    temp2: safeXml`
      <fill>
        ,
        <patternFill patternType="solid">
          <fgColor rgb="FFF2F2F2" />
          <bgColor rgb="FFC6EFCE" />
        </patternFill>
      </fill>
    `,
    temp3: safeXml`
      <fill>
        <patternFill patternType="solid">
          <fgColor rgb="FFC8F2C2" />
          <bgColor rgb="FFC6EFCE" />
        </patternFill>
      </fill>
    `,
    temp4: safeXml`
      <fill>
        <patternFill patternType="solid">
          <fgColor rgb="FFEBFCEB" />
          <bgColor rgb="FFC6EFCE" />
        </patternFill>
      </fill>
    `,
    temp5: safeXml`
      <fill>
        <patternFill patternType="solid">
          <fgColor rgb="FF000000" />
          <bgColor rgb="FFFFFFFF" />
        </patternFill>
      </fill>
    `,
  });

  const { xml: bordersXml, indexMap: borderIndexMap } = createXmlIndexMap({
    none: safeXml`
      <border>
        <left />
        <right />
        <top />
        <bottom />
        <diagonal />
      </border>
    `,
    lightGray: safeXml`
      <border>
        <left style="thin">
          <color rgb="FFD4D4D4" />
        </left>
        <right style="thin">
          <color rgb="FFD4D4D4" />
        </right>
        <top style="thin">
          <color rgb="FFD4D4D4" />
        </top>
        <bottom style="thin">
          <color rgb="FFD4D4D4" />
        </bottom>
        <diagonal />
      </border>
    `,
  });

  const { xml: cellStylesXml, indexMap: cellStyleIndexMap } = createXmlIndexMap(
    {
      temp0: safeXml`
        <xf
          numFmtId="0"
          fontId="${fontIndexMap.arial12}"
          fillId="${fillIndexMap.none}"
          borderId="${borderIndexMap.lightGray}"
          xfId="0"
          applyBorder="0"
        />
      `,
      temp1: safeXml`
        <xf
          numFmtId="0"
          fontId="${fontIndexMap.arialBold24}"
          fillId="${fillIndexMap.temp2}"
          borderId="${borderIndexMap.lightGray}"
          xfId="0"
          applyBorder="0"
        >
          <alignment wrapText="1" vertical="top" />
        </xf>
      `,
      temp2: safeXml`
        <xf
          numFmtId="0"
          fontId="${fontIndexMap.arialBold14}"
          fillId="${fillIndexMap.temp3}"
          borderId="${borderIndexMap.lightGray}"
          xfId="0"
          applyBorder="0"
        >
          <alignment wrapText="1" vertical="top" />
        </xf>
      `,
      temp3: safeXml`
        <xf
          numFmtId="0"
          fontId="${fontIndexMap.arialBold14}"
          fillId="${fillIndexMap.temp3}"
          borderId="${borderIndexMap.lightGray}"
          xfId="0"
          applyBorder="0"
        >
          <alignment wrapText="1" vertical="top" />
        </xf>
      `,
      temp4: safeXml`
        <xf
          numFmtId="0"
          fontId="${fontIndexMap.arial12}"
          fillId="${fillIndexMap.temp4}"
          borderId="${borderIndexMap.lightGray}"
          xfId="0"
          applyBorder="0"
        >
          <alignment wrapText="1" vertical="center" />
        </xf>
      `,
      temp5: safeXml`
        <xf
          numFmtId="0"
          fontId="${fontIndexMap.arialBold22}"
          fillId="${fillIndexMap.temp5}"
          borderId="${borderIndexMap.lightGray}"
          xfId="0"
          applyBorder="0"
        >
          <alignment wrapText="1" horizontal="center" vertical="center" />
        </xf>
      `,
    },
  );

  const styleXml = safeXml`
    <?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
    <styleSheet
      xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      mc:Ignorable="x14ac x16r2 xr"
      xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac"
      xmlns:x16r2="http://schemas.microsoft.com/office/spreadsheetml/2015/02/main"
      xmlns:xr="http://schemas.microsoft.com/office/spreadsheetml/2014/revision"
    >
      <fonts count="1" x14ac:knownFonts="1">${fontsXml}</fonts>
      <fills count="3">${fillsXml}</fills>
      <borders count="1">${bordersXml}</borders>
      <cellStyleXfs count="1">
        <xf numFmtId="0" fontId="0" fillId="0" borderId="1" />
      </cellStyleXfs>
      <cellXfs count="1">${cellStylesXml}</cellXfs>
      <cellStyles count="1">
        <cellStyle name="Normal" xfId="0" builtinId="0" />
      </cellStyles>
      <dxfs count="0" />
      <tableStyles
        count="0"
        defaultTableStyle="TableStyleMedium2"
        defaultPivotStyle="PivotStyleLight16"
      />
      <extLst>
        <ext
          uri="{EB79DEF2-80B8-43e5-95BD-54CBDDF9020C}"
          xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main"
        >
          <x14:slicerStyles defaultSlicerStyle="SlicerStyleLight1" />
        </ext>
        <ext
          uri="{9260A510-F301-46a8-8635-F512D64BE5F5}"
          xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
        >
          <x15:timelineStyles defaultTimelineStyle="TimeSlicerStyleLight1" />
        </ext>
      </extLst>
    </styleSheet>
  `.trim();

  return {
    styleXml,
    cellStyleIndexMap,
  };
}

function buildNatCurric<T extends Record<string, string>>(
  cellStyleIndexMap: T,
  data: BuildNationalCurriculumData,
) {
  return safeXml`
    <worksheet
      xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
      xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
      mc:Ignorable="x14ac xr xr2 xr3"
      xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac"
      xmlns:xr="http://schemas.microsoft.com/office/spreadsheetml/2014/revision"
      xmlns:xr2="http://schemas.microsoft.com/office/spreadsheetml/2015/revision2"
      xmlns:xr3="http://schemas.microsoft.com/office/spreadsheetml/2016/revision3"
      xr:uid="{BBB25A11-9FE4-904C-97A9-5616457FC1B3}"
    >
      <dimension ref="A1" />
      <sheetViews>
        <sheetView tabSelected="1" workbookViewId="0">
          <pane
            xSplit="1"
            ySplit="3"
            topLeftCell="B4"
            activePane="bottomLeft"
            state="frozen"
          />
          <selection pane="bottomLeft" activeCell="A1" sqref="A1" />
        </sheetView>
      </sheetViews>
      <sheetFormatPr
        baseColWidth="10"
        defaultRowHeight="16"
        x14ac:dyDescent="0.2"
      />
      <cols>
        <col
          min="1"
          max="2"
          width="${pxToColumnWidth(285, 7)}"
          customWidth="1"
        />
        <col
          min="2"
          max="${data.unitData.length + 1}"
          width="${pxToColumnWidth(185, 7)}"
          customWidth="1"
        />
      </cols>
      <sheetData>
        <row r="1" spans="1:26">
          <c r="A1" t="inlineStr" s="${cellStyleIndexMap.temp1!}">
            <is>
              <t>
                ${cdata(
                  `Year ${data.unitData[0]?.unit.year ?? ""} ${data.unitData[0]!.unit.subject}`,
                )}
              </t>
            </is>
          </c>
        </row>
        <row r="2" spans="1:26">
          <c r="A2" t="inlineStr" s="${cellStyleIndexMap.temp2!}">
            <is>
              <t xml:space="preserve">${cdata(
                  `\nNational curriculum statement\n`,
                )}</t>
            </is>
          </c>
          ${data.unitData.map((unit, unitIndex) => {
            return safeXml`
              <c
                r="${cartesianToExcelCoords([unitIndex + 2, 2])}"
                t="inlineStr"
                s="${cellStyleIndexMap.temp3!}"
              >
                <is>
                  <r>
                    <rPr>
                      <sz val="14" />
                      <color rgb="FF000000" />
                      <rFont val="Arial Bold" />
                      <family val="2" />
                    </rPr>
                    <t xml:space="preserve">${cdata(
                        `\nUnit ${unitIndex + 1}\n${unit.unit.title}\n`,
                      )}</t>
                  </r>
                </is>
              </c>
            `;
          })}
        </row>
        <row r="3" spans="1:26">
          <c r="A3" t="inlineStr" s="${cellStyleIndexMap.temp2!}">
            <is>
              <t />
            </is>
          </c>
          ${data.unitData.map((_unit, unitIndex) => {
            return safeXml`
              <c
                r="${cartesianToExcelCoords([unitIndex + 2, 3])}"
                t="inlineStr"
                s="${cellStyleIndexMap.temp3!}"
              >
                <is>
                  <r>
                    <rPr>
                      <sz val="12" />
                      <color rgb="FF000000" />
                      <rFont val="Arial" />
                      <family val="2" />
                      <u val="single" />
                    </rPr>
                    <t>Go to unit resources</t>
                  </r>
                </is>
              </c>
            `;
          })}
        </row>
        ${[...data.nationalCurric.entries()].map(
          ([id, nationalCurricText], nationalCurricTextIndex) => {
            const yPos = 4 + nationalCurricTextIndex;
            return safeXml`
              <row r="${yPos}" spans="1:${data.unitData.length + 1}">
                <c
                  r="${cartesianToExcelCoords([1, yPos])}"
                  t="inlineStr"
                  s="${cellStyleIndexMap.temp4!}"
                >
                  <is>
                    <t xml:space="preserve">${cdata(
                        `\n${nationalCurricText}\n`,
                      )}</t>
                  </is>
                </c>
                ${data.unitData.map((unit, unitIndex) => {
                  return safeXml`
                    <c
                      r="${cartesianToExcelCoords([unitIndex + 2, yPos])}"
                      t="inlineStr"
                      s="${unit.nationalCurricIds.includes(id)
                        ? cellStyleIndexMap.temp5!
                        : cellStyleIndexMap.temp0!}"
                    >
                      <is>
                        <t>
                          ${cdata(
                            unit.nationalCurricIds.includes(id) ? "✓" : "",
                          )}
                        </t>
                      </is>
                    </c>
                  `;
                })}
              </row>
            `;
          },
        )}
      </sheetData>
      <hyperlinks>
        ${data.unitData.map((unit, unitIndex) => {
          return safeXml`
            <hyperlink
              ref="${cartesianToExcelCoords([unitIndex + 2, 3])}"
              r:id="rId${1000 + unitIndex}"
            />
          `;
        })}
      </hyperlinks>
      <pageMargins
        left="0.7"
        right="0.7"
        top="0.75"
        bottom="0.75"
        header="0.3"
        footer="0.3"
      />
    </worksheet>
  `.trim();
}

type BuildNationalCurriculumData = {
  year: string;
  nationalCurric: Map<number, string>;
  unitData: {
    unit: Unit;
    nationalCurricIds: number[];
  }[];
};
async function buildNationalCurriculum(
  zip: JSZipCached,
  data: BuildNationalCurriculumData[],
) {
  const { styleXml, cellStyleIndexMap } = buildStyle();
  zip.writeString("xl/styles.xml", styleXml);

  data.forEach((item, index) => {
    zip.writeString(
      `xl/worksheets/_rels/sheet${10 + index}.xml.rels`,
      safeXml`
        <?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
        <Relationships
          xmlns="http://schemas.openxmlformats.org/package/2006/relationships"
        >
          ${item.unitData.map((unit, unitIndex) => {
            return safeXml`
              <Relationship
                Id="rId${1000 + unitIndex}"
                Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink"
                Target="https://www.thenational.academy/teachers/curriculum/${unit
                  .unit.subject_slug}-${unit.unit.phase_slug}/units/${unit.unit
                  .slug}"
                TargetMode="External"
              />
            `;
          })}
        </Relationships>
      `.trim(),
    );
    addOrUpdateSheet(
      zip,
      10 + index,
      xmlCompact(buildNatCurric(cellStyleIndexMap, item)),
    );
  });

  zip.writeString(
    "xl/_rels/workbook.xml.rels",
    safeXml`
      <?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
      <Relationships
        xmlns="http://schemas.openxmlformats.org/package/2006/relationships"
      >
        <Relationship
          Id="rId3"
          Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles"
          Target="styles.xml"
        />
        <Relationship
          Id="rId2"
          Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/theme"
          Target="theme/theme1.xml"
        />
        ${data.map((_, index) => {
          return safeXml`
            <Relationship
              Id="rId${10 + index}"
              Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet"
              Target="worksheets/sheet${10 + index}.xml"
            />
          `;
        })}
      </Relationships>
    `.trim(),
  );

  zip.writeString(
    "xl/workbook.xml",
    safeXml`
      <?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
      <workbook
        xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
        xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="x15 xr xr6 xr10 xr2"
        xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
        xmlns:xr="http://schemas.microsoft.com/office/spreadsheetml/2014/revision"
        xmlns:xr6="http://schemas.microsoft.com/office/spreadsheetml/2016/revision6"
        xmlns:xr10="http://schemas.microsoft.com/office/spreadsheetml/2016/revision10"
        xmlns:xr2="http://schemas.microsoft.com/office/spreadsheetml/2015/revision2"
      >
        <fileVersion
          appName="xl"
          lastEdited="7"
          lowestEdited="7"
          rupBuild="10511"
        />
        <workbookPr defaultThemeVersion="202300" />
        <mc:AlternateContent
          xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        >
          <mc:Choice Requires="x15">
            <x15ac:absPath
              url="/Users/orangemug/Documents/"
              xmlns:x15ac="http://schemas.microsoft.com/office/spreadsheetml/2010/11/ac"
            />
          </mc:Choice>
        </mc:AlternateContent>
        <xr:revisionPtr
          revIDLastSave="0"
          documentId="8_{274615AA-ECE0-644D-AB2E-12D373184771}"
          xr6:coauthVersionLast="47"
          xr6:coauthVersionMax="47"
          xr10:uidLastSave="{00000000-0000-0000-0000-000000000000}"
        />
        <bookViews>
          <workbookView
            xWindow="1100"
            yWindow="820"
            windowWidth="28040"
            windowHeight="17440"
            xr2:uid="{C98B37CC-FD85-FE4F-984F-54449C3C0043}"
          />
        </bookViews>
        <sheets>
          ${data.map((item, index) => {
            return safeXml`
              <sheet
                name="Year ${item.year}"
                sheetId="${index + 1}"
                r:id="rId${10 + index}"
              />
            `;
          })}
        </sheets>
        <calcPr calcId="181029" />
        <extLst>
          <ext
            uri="{140A7094-0E35-4892-8432-C4D2E57EDEB5}"
            xmlns:x15="http://schemas.microsoft.com/office/spreadsheetml/2010/11/main"
          >
            <x15:workbookPr chartTrackingRefBase="1" />
          </ext>
          <ext
            uri="{B58B0392-4F1F-4190-BB64-5DF3571DCE5F}"
            xmlns:xcalcf="http://schemas.microsoft.com/office/spreadsheetml/2018/calcfeatures"
          >
            <xcalcf:calcFeatures>
              <xcalcf:feature name="microsoft.com:RD" />
              <xcalcf:feature name="microsoft.com:Single" />
              <xcalcf:feature name="microsoft.com:FV" />
              <xcalcf:feature name="microsoft.com:CNMTM" />
              <xcalcf:feature name="microsoft.com:LET_WF" />
              <xcalcf:feature name="microsoft.com:LAMBDA_WF" />
              <xcalcf:feature name="microsoft.com:ARRAYTEXT_WF" />
            </xcalcf:calcFeatures>
          </ext>
        </extLst>
      </workbook>
    `.trim(),
  );
}

export default async function xlsxNationalCurriculum(
  data: CurriculumUnitsTabData & CurriculumOverviewMVData,
  // slugs: Slugs,
  // ks4Options: Ks4Option[],
) {
  const zip = await generateEmptyXlsx();

  const formattedData = formatCurriculumUnitsData(data);

  const obj = Object.entries(formattedData.yearData).map(
    ([year, { units }]) => {
      const nationalCurric = new Map<number, string>();

      const unitData = units.map((unit) => {
        const ids: number[] = [];
        unit.national_curriculum_content?.forEach((cc) => {
          nationalCurric.set(cc.id, cc.title);
          ids.push(cc.id);
        });

        return {
          unit,
          nationalCurricIds: ids,
        };
      });

      return {
        year,
        nationalCurric,
        unitData,
      };
    },
  );

  await buildNationalCurriculum(zip, obj);

  return await zip.zipToBuffer();
}
