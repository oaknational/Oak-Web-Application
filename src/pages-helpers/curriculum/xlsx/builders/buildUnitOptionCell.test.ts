import { prettyFormat } from "../../docx/xml";

import { buildUnitOptionCell } from "./buildUnitOptionCell";

import { createUnitOption } from "@/fixtures/curriculum/unitOption";

test("buildUnitOptionCell", () => {
  const output = buildUnitOptionCell({
    cellStyleIndexMap: { temp3: "TEMP_3" },
    x: 1,
    y: 1,
    unitOption: createUnitOption(),
    unitOptionIndex: 1,
  });
  expect(prettyFormat(output)).toMatchSnapshot();
});
