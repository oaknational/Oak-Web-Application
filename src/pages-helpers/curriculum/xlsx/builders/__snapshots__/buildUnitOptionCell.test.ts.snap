// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildUnitOptionCell 1`] = `
"<c
  r="A1"
  t="inlineStr"
  s="TEMP_3"
>
  <is>
    <r>
      <rPr>
        <sz
          val="14"
        ></sz>
        <color
          rgb="FF000000"
        ></color>
        <rFont
          val="Arial"
        ></rFont>
        <family
          val="2"
        ></family>
      </rPr>
      <t
        xml:space="preserve"
      ><![CDATA[
Option 2
]]></t>
    </r>
    <r>
      <rPr>
        <sz
          val="14"
        ></sz>
        <color
          rgb="FF000000"
        ></color>
        <rFont
          val="Arial Bold"
        ></rFont>
        <family
          val="2"
        ></family>
      </rPr>
      <t
        xml:space="preserve"
      ><![CDATA[Test
]]></t>
    </r>
  </is>
</c>"
`;
